src/speculators/models/llama_eagle.py

```diff
@@ -0,0 +1,276 @@
+    )
+
+
+class EagleLlamaDecoderLayer(HFLlamaDecoderLayer):
+    """
+    A LlamaDecoderLayer variant for the Eagle speculator, allowing the input
+    LayerNorm to be disabled, typically for the first layer.
+    """
+    def __init__(self, config: LlamaConfig, layer_idx: int, disable_input_layernorm: bool = True):
+        super().__init__(config, layer_idx)
+        if disable_input_layernorm and hasattr(self, 'input_layernorm') and \
+           isinstance(self.input_layernorm, (LlamaRMSNorm, nn.LayerNorm)):
+            self.input_layernorm = nn.Identity()
+
+
+# TODO: Should LlamaEagleSpeculator inherit from AutoModel?
+class LlamaEagleSpeculator(nn.Module):
+    """
+    Llama-based Eagle speculator model.
+
+    This model architecture is designed to generate speculative tokens based on
+    inputs from a verifier model and its own previous predictions. It features
+    configurable input fusion and optional LayerNorm stages.
+    """
+    def __init__(self, config: LlamaEagleConfig):
+        super().__init__()
+        self.config = config
+        self.speculators_config = config.speculators_config
+        self.verifier_config = self.speculators_config.verifier
+
+        self._setup_core_properties()
+        self._setup_embeddings()
+        self._setup_optional_input_layernorms()
+
+        # Speculator's own embedding dimension
+        # Determine the hidden size for the main transformer layers, which might change
+        # based on the input(s)
+        speculator_embed_dim = self.config.hidden_size
+        current_transformer_hidden_size = self.config.hidden_size
+
+        self.input_projector, self.post_projection_act, speculator_transformer_hidden_size = \
+            self._setup_input_fusion(speculator_embed_dim, current_transformer_hidden_size)
+
+        self._setup_transformer_stack(speculator_transformer_hidden_size)
+        self._setup_output_processing(speculator_transformer_hidden_size)
+        self._apply_target_dtype()
+
+    def _setup_core_properties(self):
+        """Sets up core model properties like gradient checkpointing."""
+        self.gradient_checkpointing = getattr(self.config, "gradient_checkpointing", False)
+        self.padding_idx = getattr(self.config, "pad_token_id", None) # Typically the pad_token_id from verifier config???
+
+    def _setup_embeddings(self):
+        """Initializes and optionally freezes the token embedding layer."""
+        speculator_embed_dim = self.config.hidden_size
+        self.embed_tokens = nn.Embedding(
+            self.verifier_config.vocab_size,
+            speculator_embed_dim,
+            self.padding_idx
+        )
+        if getattr(self.config, "freeze_embed_tokens", True):
+            self.embed_tokens.requires_grad_(False)
+
+    def _setup_optional_input_layernorms(self):
+        """
+        Initializes optional LayerNorms for inputs, based on config flags.
+        These were present in the HASS/TTT model.
```

Comment:
typo: I meant this model https://huggingface.co/nm-testing/Eagle_Speculator_Llama_3_1_8B_TTT/tree/main